[Root Signatures] Updating docs with Version 1.2 additions (#342)
llvm/wg-hlsl: proposals/0002-root-signature-in-clang.md

    STATIC_BORDER_COLOR = 'STATIC_BORDER_COLOR_TRANSPARENT_BLACK' |
                          'STATIC_BORDER_COLOR_OPAQUE_BLACK' |
                          'STATIC_BORDER_COLOR_OPAQUE_WHITE' |
                          'STATIC_BORDER_COLOR_OPAQUE_BLACK_UINT' |
                          'STATIC_BORDER_COLOR_OPAQUE_WHITE_UINT' ;

    STATIC_SAMPLER_FLAGS = 'SAMPLER_FLAG_UINT_BORDER_COLOR' |
                           'SAMPLER_FLAG_NON_NORMALIZED_COORDINATES';
```

### Root Signature Versioning 

> Currently, DirectX supports three "versions" of root signatures: 1.0, 1.1 and 1.2. 

Reviewer: nit: Why is versions in quotes?

Author: Root Signatures "versions" are not a redesign, or refactoring, or changes replacing previous versions, they are adding new features on top of existing ones. So, all root signatures that can be written in version 1.0 are still valid in version 1.1. I believe that was the intent when using quotes here. Happy to remove it, if folks prefer that way.

Version 1.1 includes additional flags for descriptor ranges and root descriptors.
See the [DirectX Documentation][root_signature_versions_doc] for full details.
Version 1.2 includes additions flags for static samplers. This changes were documented

> in the [Vulkan Specification][root_signature_version12_doc] for root signatures.

Reviewer 2: This isn't the "Vulkan Specification" - this is the spec for VulkanOn12 (that is, the implementation of Vulkan on D3D12)

    STATIC_BORDER_COLOR = 'STATIC_BORDER_COLOR_TRANSPARENT_BLACK' |
                          'STATIC_BORDER_COLOR_OPAQUE_BLACK' |
                          'STATIC_BORDER_COLOR_OPAQUE_WHITE' ;

>     STATIC_SAMPLER_FLAGS = 'SAMPLER_FLAG_UINT_BORDER_COLOR' |

Author: This flag is used by the vulkan backend, although we don't support it yet, it should be added as part of the current work

    - `SRV`: `DATA_VOLATILE | DESCRIPTORS_VOLATILE`
    - `UAV`: `DATA_VOLATILE | DESCRIPTORS_VOLATILE`
    - `Sampler`: `DESCRIPTORS_VOLATILE`

>   - Version 1.1 onwards:

Reviewer: ```suggestion
  - Version 1.1 and onwards:
```

  Note that a valid value for `numDescriptors` is `unbounded` and requires
  overlap analysis.

- When flag `SAMPLER_FLAG_NON_NORMALIZED_COORDINATES` is set:
  - `Filter` must be one of
    - `FILTER_MIN_MAG_MIP_POINT`
    - `FILTER_MIN_MAG_LINEAR_MIP_POINT`
    - `FILTER_MINIMUM_MIN_MAG_MIP_POINT`
    - `FILTER_MINIMUM_MIN_MAG_LINEAR_MIP_POINT`
    - `FILTER_MAXIMUM_MIN_MAG_MIP_POINT`
    - `FILTER_MAXIMUM_MIN_MAG_LINEAR_MIP_POINT`
  - `MinLOD` and `MaxLOD` must be 0
  - `AddressU` and `AddressV` must be `TEXTURE_ADDRESS_MODE_CLAMP` or `TEXTURE_ADDRESS_MODE_BORDER`

> - When flag `SAMPLER_FLAG_UINT_BORDER_COLOR` is set:

Reviewer: nit:
```suggestion
- When the flag `SAMPLER_FLAG_UINT_BORDER_COLOR` is set:
```

* i32: ShaderRegister
* i32: RegisterSpace
* i32: ShaderVisibility ([D3D12_SHADER_VISIBILITY][d3d12_shader_visibility])
* i32: Flags ([D3D12_SAMPLER_FLAGS][root_signature_version12_doc])

[root_signature_versions_doc]: https://learn.microsoft.com/en-us/windows/win32/direct3d12/root-signature-version-1-1

> [root_signature_version12_doc]: https://microsoft.github.io/DirectX-Specs/d3d/VulkanOn12.html#definition-9

Reviewer 2: It doesn't matter much since the reference name isn't really visible anywhere, but we might as well call this `root_signature_vulkanon12` for clarity, since this isn't actually a document about root signature 1.2 like the 1.1 doc above is.

    When the Filter of a StaticSampler is `FILTER_COMPARISON*`,
    the ComparisonFunc cannot be 0.
  
  - Valid values for Flags
    - For root signature version 1.0 and 1.1 must be 0
    - For root signature version 1.2:

>       - SAMPLER_FLAG_UINT_BORDER_COLOR

Reviewer 3: When SAMPLER_FLAG_UINT_BORDER_COLOR is set the border color must be on eof:

* opaque black uint
* opaque white uint
* transparent black

I think that this also suggests that we need to add these colors as well because they're not currently listed in the grammar section of this doc at least.  See https://github.com/microsoft/DirectX-Headers/blob/b8238e686c4f8acfe234e9800e89612444f7b758/include/directx/d3d12.h#L4087

    When the Filter of a StaticSampler is `FILTER_COMPARISON*`,
    the ComparisonFunc cannot be 0.
  
  - Valid values for Flags

>     - For root signature version 1.0 and 1.1 must be 0

Reviewer: nit
```suggestion
    - For root signature versions 1.0 and 1.1, Flags must be 0
```
